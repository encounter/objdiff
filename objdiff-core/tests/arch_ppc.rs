use objdiff_core::{
    diff::{self, display},
    obj,
    obj::SectionKind,
};

mod common;

#[test]
#[cfg(feature = "ppc")]
fn read_ppc() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/ppc/IObj.o"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx =
        obj.symbols.iter().position(|s| s.name == "Type2Text__10SObjectTagFUi").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}

#[test]
#[cfg(feature = "ppc")]
fn read_dwarf1_line_info() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/ppc/m_Do_hostIO.o"), &diff_config).unwrap();
    let line_infos = obj
        .sections
        .iter()
        .filter(|s| s.kind == SectionKind::Code)
        .map(|s| s.line_info.clone())
        .collect::<Vec<_>>();
    insta::assert_debug_snapshot!(line_infos);
}

#[test]
#[cfg(feature = "ppc")]
fn read_extab() {
    let diff_config = diff::DiffObjConfig::default();
    let obj = obj::read::parse(include_object!("data/ppc/NMWException.o"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj);
}

#[test]
#[cfg(feature = "ppc")]
fn diff_ppc() {
    let diff_config = diff::DiffObjConfig::default();
    let mapping_config = diff::MappingConfig::default();
    let target_obj =
        obj::read::parse(include_object!("data/ppc/CDamageVulnerability_target.o"), &diff_config)
            .unwrap();
    let base_obj =
        obj::read::parse(include_object!("data/ppc/CDamageVulnerability_base.o"), &diff_config)
            .unwrap();
    let diff =
        diff::diff_objs(Some(&target_obj), Some(&base_obj), None, &diff_config, &mapping_config)
            .unwrap();

    let target_diff = diff.left.as_ref().unwrap();
    let base_diff = diff.right.as_ref().unwrap();
    let sections_display = display::display_sections(
        &target_obj,
        target_diff,
        display::SymbolFilter::None,
        false,
        false,
        true,
    );
    insta::assert_debug_snapshot!(sections_display);

    let target_symbol_idx = target_obj
        .symbols
        .iter()
        .position(|s| s.name == "WeaponHurts__20CDamageVulnerabilityCFRC11CWeaponModei")
        .unwrap();
    let target_symbol_diff = &target_diff.symbols[target_symbol_idx];
    let base_symbol_idx = base_obj
        .symbols
        .iter()
        .position(|s| s.name == "WeaponHurts__20CDamageVulnerabilityCFRC11CWeaponModei")
        .unwrap();
    let base_symbol_diff = &base_diff.symbols[base_symbol_idx];
    assert_eq!(target_symbol_diff.target_symbol, Some(base_symbol_idx));
    assert_eq!(base_symbol_diff.target_symbol, Some(target_symbol_idx));
    insta::assert_debug_snapshot!((target_symbol_diff, base_symbol_diff));
}

#[test]
#[cfg(feature = "ppc")]
fn read_vmx128_coff() {
    let diff_config = diff::DiffObjConfig { combine_data_sections: true, ..Default::default() };
    let obj = obj::read::parse(include_object!("data/ppc/vmx128.obj"), &diff_config).unwrap();
    insta::assert_debug_snapshot!(obj);
    let symbol_idx =
        obj.symbols.iter().position(|s| s.name == "?FloatingPointExample@@YAXXZ").unwrap();
    let diff = diff::code::no_diff_code(&obj, symbol_idx, &diff_config).unwrap();
    insta::assert_debug_snapshot!(diff.instruction_rows);
    let output = common::display_diff(&obj, &diff, symbol_idx, &diff_config);
    insta::assert_snapshot!(output);
}
