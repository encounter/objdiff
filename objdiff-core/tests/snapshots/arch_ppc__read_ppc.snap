---
source: objdiff-core/tests/arch_ppc.rs
assertion_line: 14
expression: obj
---
Object {
    arch: ArchPpc {
        extab: None,
    },
    endianness: Big,
    symbols: [
        Symbol {
            name: "IObj.cpp",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Local),
            align: None,
            virtual_address: Some(
                0,
            ),
        },
        Symbol {
            name: "[.text]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                0,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: Some(
                2150895620,
            ),
        },
        Symbol {
            name: "[.ctors]",
            demangled_name: None,
            address: 0,
            size: 4,
            kind: Section,
            section: Some(
                1,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: Some(
                2151461704,
            ),
        },
        Symbol {
            name: "[.sbss]",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Section,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: Some(
                2153420048,
            ),
        },
        Symbol {
            name: "__sinit_IObj_cpp",
            demangled_name: None,
            address: 264,
            size: 20,
            kind: Function,
            section: Some(
                0,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: Some(
                2150895884,
            ),
        },
        Symbol {
            name: "text$52",
            demangled_name: None,
            address: 8,
            size: 5,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Local),
            align: None,
            virtual_address: Some(
                2153420056,
            ),
        },
        Symbol {
            name: "Type2Text__10SObjectTagFUi",
            demangled_name: Some(
                "SObjectTag::Type2Text(unsigned int)",
            ),
            address: 0,
            size: 264,
            kind: Function,
            section: Some(
                0,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: Some(
                2150895620,
            ),
        },
        Symbol {
            name: "gkInvalidObjectTag",
            demangled_name: None,
            address: 0,
            size: 8,
            kind: Object,
            section: Some(
                2,
            ),
            flags: FlagSet(Global),
            align: None,
            virtual_address: Some(
                2153420048,
            ),
        },
        Symbol {
            name: "__upper_map",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: Some(
                0,
            ),
        },
        Symbol {
            name: "__ctype_map",
            demangled_name: None,
            address: 0,
            size: 0,
            kind: Unknown,
            section: None,
            flags: FlagSet(Global),
            align: None,
            virtual_address: Some(
                0,
            ),
        },
    ],
    sections: [
        Section {
            id: ".text-0",
            name: ".text",
            address: 0,
            size: 284,
            kind: Code,
            data: SectionData(
                284,
            ),
            flags: FlagSet(),
            relocations: [
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 22,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 26,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 28,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 36,
                    target_symbol: 5,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 58,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 62,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 64,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 72,
                    target_symbol: 5,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 98,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 102,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 104,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 112,
                    target_symbol: 5,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 138,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 142,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 144,
                    target_symbol: 8,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 148,
                    target_symbol: 5,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        6,
                    ),
                    address: 162,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        4,
                    ),
                    address: 166,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 176,
                    target_symbol: 5,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 180,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 200,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 220,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        0,
                    ),
                    address: 240,
                    target_symbol: 9,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 256,
                    target_symbol: 5,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 268,
                    target_symbol: 7,
                    addend: 0,
                },
                Relocation {
                    flags: Elf(
                        109,
                    ),
                    address: 272,
                    target_symbol: 7,
                    addend: 0,
                },
            ],
            line_info: {},
            virtual_address: Some(
                2150895620,
            ),
        },
        Section {
            id: ".ctors-0",
            name: ".ctors",
            address: 0,
            size: 4,
            kind: Data,
            data: SectionData(
                4,
            ),
            flags: FlagSet(),
            relocations: [
                Relocation {
                    flags: Elf(
                        1,
                    ),
                    address: 0,
                    target_symbol: 4,
                    addend: 0,
                },
            ],
            line_info: {},
            virtual_address: Some(
                2151461704,
            ),
        },
        Section {
            id: ".sbss-0",
            name: ".sbss",
            address: 0,
            size: 16,
            kind: Bss,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: Some(
                2153420048,
            ),
        },
        Section {
            id: ".rela.text-0",
            name: ".rela.text",
            address: 0,
            size: 216,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".rela.ctors-0",
            name: ".rela.ctors",
            address: 0,
            size: 12,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".symtab-0",
            name: ".symtab",
            address: 0,
            size: 176,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".strtab-0",
            name: ".strtab",
            address: 0,
            size: 105,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".shstrtab-0",
            name: ".shstrtab",
            address: 0,
            size: 77,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".comment-0",
            name: ".comment",
            address: 0,
            size: 132,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
        Section {
            id: ".note.split-0",
            name: ".note.split",
            address: 0,
            size: 152,
            kind: Unknown,
            data: SectionData(
                0,
            ),
            flags: FlagSet(),
            relocations: [],
            line_info: {},
            virtual_address: None,
        },
    ],
    split_meta: Some(
        SplitMeta {
            generator: Some(
                "decomp-toolkit 1.4.0",
            ),
            module_name: Some(
                "main",
            ),
            module_id: Some(
                0,
            ),
            virtual_addresses: Some(
                [
                    0,
                    0,
                    2150895620,
                    2151461704,
                    2153420048,
                    2150895884,
                    2153420056,
                    2150895620,
                    2153420048,
                    0,
                    0,
                ],
            ),
        },
    ),
    path: None,
    timestamp: None,
}
